### PR TITLE
tutorial example for binary grating at oblique incidence

### LINKED ISSUE #486: reflectance and transmittance of diffracted orders of binary grating for oblique source don't sum to one
As a validation of the mode-decomposition feature, which could also possibly be added to the test suite, the reflectance and transmittance of all diffracted orders for any non-lossy grating structure for an incident planewave source at any arbitrary angle must necessarily sum to one. The following script, which is an adaptation of the [binary grating tutorial example](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#diffraction-spectrum-of-a-binary-grating), sets up this test. For comparison, the script also computes the reflectance and transmittance directly via the Poynting flux. The total reflectance, transmittance, and their sum are displayed at the end of the simulation in two separate lines prefixed by `coeff:` (mode coefficients) and `flux:` (Poynting flux).

```py
mport meep as mp
import math
import cmath
import numpy as np

resolution = 40        # pixels/μm                                                                                                                                                              

dpml = 2.0             # PML thickness                                                                                                                                                          
dsub = 3.0             # substrate thickness                                                                                                                                                    
dpad = 3.0             # length of padding between grating and pml                                                                                                                              
gp = 10.0              # grating period                                                                                                                                                         
gh = 0.5               # grating height                                                                                                                                                         
gdc = 0.5              # grating duty cycle                                                                                                                                                     

sx = dpml+dsub+gh+dpad+dpml
sy = gp

cell_size = mp.Vector3(sx,sy,0)

# replace anisotropic PML with isotropic Absorber to attenuate parallel-directed fields of oblique source                                                                                       
abs_layers = [mp.Absorber(thickness=dpml,direction=mp.X)]

wvl = 0.5              # center wavelength                                                                                                                                                      
fcen = 1/wvl           # center frequency                                                                                                                                                       
df = 0.05*fcen         # frequency width                                                                                                                                                        

ng = 1.5
glass = mp.Medium(index=ng)

# rotation angle of incident planewave; CCW about Y axis, 0 degrees along +X axis                                                                                                                        
theta_in = math.radians(10.7)

# k (in vacuum) with correct length (plane of incidence: XY)                                                                                                                                    
k = mp.Vector3(math.cos(theta_in),math.sin(theta_in),0).scale(fcen)

symmetries = []
eig_parity = mp.ODD_Z
if theta_in == 0:
  k = mp.Vector3(0,0,0)
  symmetries = [mp.Mirror(mp.Y)]
  eig_parity += mp.EVEN_Y

def pw_amp(k,x0):
  def _pw_amp(x):
    return cmath.exp(1j*2*math.pi*ng*k.dot(x+x0))
  return _pw_amp

src_pt = mp.Vector3(-0.5*sx+dpml+0.3*dsub,0,0)
sources = [mp.Source(mp.GaussianSource(fcen,fwidth=df),
                     component=mp.Ez,
                     center=src_pt,
                     size=mp.Vector3(0,sy,0),
                     amp_func=pw_amp(k,src_pt))]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=abs_layers,
                    k_point=k,
                    default_material=glass,
                    sources=sources,
                    symmetries=symmetries)

refl_pt = mp.Vector3(-0.5*sx+dpml+0.5*dsub,0,0)
refl_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=refl_pt, size=mp.Vector3(0,sy,0)))

sim.run(until_after_sources=1000)

input_flux = mp.get_fluxes(refl_flux)
input_flux_data = sim.get_flux_data(refl_flux)

sim.reset_meep()

geometry = [mp.Block(material=glass, size=mp.Vector3(dpml+dsub,mp.inf,mp.inf), center=mp.Vector3(-0.5*sx+0.5*(dpml+dsub),0,0)),
            mp.Block(material=glass, size=mp.Vector3(gh,gdc*gp,mp.inf), center=mp.Vector3(-0.5*sx+dpml+dsub+0.5*gh,0,0))]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=abs_layers,
                    geometry=geometry,
                    k_point=k,
                    sources=sources,
                    symmetries=symmetries)

refl_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=refl_pt, size=mp.Vector3(0,sy,0)))
sim.load_minus_flux_data(refl_flux,input_flux_data)

tran_pt = mp.Vector3(0.5*sx-dpml-0.5*dpad,0,0)
tran_flux = sim.add_flux(fcen, 0, 1, mp.FluxRegion(center=tran_pt, size=mp.Vector3(0,sy,0)))

sim.run(until_after_sources=mp.stop_when_fields_decayed(50, mp.Ez, src_pt, 1e-7))

r_flux = mp.get_fluxes(refl_flux)
t_flux = mp.get_fluxes(tran_flux)

r_nmode = np.floor((fcen*ng-k.y)*gp)-np.ceil((-fcen*ng-k.y)*gp) # number of reflected orders                                                                                                    
if theta_in == 0:
  r_nmode = r_nmode*0.5 # since eig_parity=mp.ODD_Z+mp.EVEN_Y which removes degeneracy                                                                                                          
Rsum = 0
for nm in range(int(r_nmode)):
  r_coeffs, r_vgrps, r_kpoints = sim.get_eigenmode_coefficients(refl_flux, [nm+1], eig_parity=eig_parity)
  Rmode = (abs(r_coeffs[0,0,1])**2)/input_flux[0]
  r_angle = math.acos(r_kpoints[0].x/(ng*fcen))
  ky = ng*fcen*math.sin(r_angle)
  if theta_in != 0:
    mnega = (ky+k.y)*gp           # mode number of negative diffracted order                                                                                                                    
    if mnega % int(mnega) < 1e-4: # determine if mode is a negative diffracted order                                                                                                            
      r_angle = -r_angle
  print("refl:, {}, {:.2f}, {:.8f}".format(nm,math.degrees(r_angle),Rmode))
  Rsum += Rmode

t_nmode = np.floor((fcen-k.y)*gp)-np.ceil((-fcen-k.y)*gp)       # number of transmitted orders                                                                                                  
if theta_in == 0:
  t_nmode = t_nmode*0.5 # since eig_parity=mp.ODD_Z+mp.EVEN_Y which removes degeneracy                                                                                                          
Tsum = 0
for nm in range(int(t_nmode)):
  t_coeffs, t_vgrps, t_kpoints = sim.get_eigenmode_coefficients(tran_flux, [nm+1], eig_parity=eig_parity)
  Tmode = (abs(t_coeffs[0,0,0])**2)/input_flux[0]
  t_angle = math.acos(t_kpoints[0].x/fcen)
  ky = fcen*math.sin(t_angle)
  if theta_in != 0:
    mnega = (ky+k.y)*gp           # mode number of negative diffracted order                                                                                                                    
    if mnega % int(mnega) < 1e-4: # determine if mode is a negative diffracted order                                                                                                            
      t_angle = -t_angle
  print("tran:, {}, {:.2f}, {:.8f}".format(nm,math.degrees(t_angle),Tmode))
  Tsum += Tmode

# display total reflectance, transmittance, and their sum of all diffracted orders computed using mode coefficients
print("coeff:, {:.6f}, {:.6f}, {:.6f}".format(Rsum,Tsum,Rsum+Tsum))

# display total reflectance, transmittance, and their sum computed directly from Poynting flux
Rflux = -r_flux[0]/input_flux[0]
Tflux =  t_flux[0]/input_flux[0]
print("flux:, {:.6f}, {:.6f}, {:.6f}".format(Rflux,Tflux,Rflux+Tflux))
```
Note that, following the discussion regarding oblique planewave sources in #471, the frequency width (`df`) is quite narrow (`0.05*fcen`) and also the `PML` has been replaced with an `Absorber`. Also, the sign of the angle of each diffracted order is determined following #476.

When the planewave is normally incident (`theta_in = 0`), the total reflectance and transmittance of all diffracted orders obtained via `get_eigenmode_coefficients` does indeed sum to one. Also, the mode-coefficient values agree with those from the corresponding Poynting flux:

```
coeff:, 0.045819, 0.926615, 0.972435
flux:, 0.048261, 0.950863, 0.999124
```
However, for an oblique source (`theta_in = 10.7`), the values are significantly different. Also, they do not  seem to be converging with increasing resolution:

**resolution=40**
```
coeff:, 0.029142, 0.623634, 0.652776
flux:, 0.056826, 0.942414, 0.999240
```

**resolution=50**
```
coeff:, 0.028733, 0.542246, 0.570979
flux:, 0.062082, 0.937214, 0.999296
```

**resolution=80**
```
coeff:, 0.022980, 0.842190, 0.865170
flux:, 0.059960, 0.940259, 1.000219
```

**resolution=100**
```
coeff:, 0.066360, 0.470629, 0.536989
flux:, 0.060384, 0.939963, 1.000347
```
Note the large change in the mode-coefficient values with resolution. Increasing both the `eig_tolerance` and `eig_resolution` parameters in the call to `get_eigenmode_coefficients` did not change the results. There seems to be another discretization effect at work; a potential bug in the implementation of `get_eigenmode_coefficients`?

### PR DESCRIPTION
**Pull request by oskooi**
Closes #486.

This should be merged after #499. We will also need to add a new test.

### CODE REVIEW

File: doc/docs/Python_Tutorials/Mode_Decomposition.md [hunk 1]
```diff
@@ -299,6 +300,184 @@ The diffraction orders/modes are a finite set of propagating planewaves. The wav
...
...t at an angle of 10.7° (`theta_in`). Its spatial profile is defined using the source amplitude function (`pw_amp`). This function takes two arguments, the wavevector and a point in space (both `mp.Vector3`s), and returns a function of one argument which returns the planewave amplitude at that point. Two modifications to the original simulation are necessary for mitigating the intrinsic discretization effects of the Yee grid for oblique planewaves: (1) the pulse bandwidth is narrowed and (2) the anisotropic `PML` is replaced with an isotropic `Absorber`. Also, the `stop_when_fields_decayed` termination criteria is replaced with `until_after_sources`. As a general rule of thumb, the more oblique the planewave source, the longer the run time required to ensure accurate results. Note that there is a second line monitor between the source and the grating for computing the reflectance. The angle of each reflected/transmitted mode, which can be positive or negative, is computed using its dominant planewave vector. Since the oblique source breaks the symmetry in the $y$ direction, each diffracted order must be computed separately. In total, there are 59 reflected and 39 transmitted orders.
```
Comment (stevengj) [inline doc/docs/Python_Tutorials/Mode_Decomposition.md:446]: Ardavan, you wrote that you narrowed the bandwidth to "mitigate discretization effects".  What are you referring to here?  The discretization error at a given frequency should be independent of the bandwidth, by linearity.

One of my students pointed out the corresponding sentence in the current manual (https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#reflectance-and-transmittance-spectra-for-planewave-at-oblique-incidence) which traces back to this PR, and it is rather confusing to me.

Comment (stevengj) [reply doc/docs/Python_Tutorials/Mode_Decomposition.md:446]: (Also it's not clear why you need to replace the PML with a scalar absorber — PML seems like it should work fine here?)

Comment (oskooi) [reply doc/docs/Python_Tutorials/Mode_Decomposition.md:446]: I was referring here to the numerical dispersion which arises for planewaves propagating at an oblique angle relative to the Cartesian Yee grid. When I was setting up this tutorial, I found that the broader the bandwidth of the pulsed source, the longer it was taking for the fields to decay away and thus for the simulation to be terminated. I think there could be some interaction of the oblique waves and the absorbing boundary layers which was causing poor attenuation. The workaround was to define a narrow-bandwidth source.

I discovered after this PR that the choice of `Absorber` and `PML` actually did not affect the results. This is why the `Absorber` was replaced with a `PML` in a [subsequent change](https://github.com/NanoComp/meep/pull/605/files#diff-0205fe8da836afa86c9d910a14fb9eab8d0fc270c086bd8db2beb087b11a7207R21) in #605.

Comment (stevengj) [reply doc/docs/Python_Tutorials/Mode_Decomposition.md:446]: This is not a discretization issue — if the bandwidth is too broad, then you are creating glancing-angle waves, which aren't absorbed well by the PML (or by an absorber).

**Pull request merged**